### DISCUSSION TITLE
Server Setting to enable CORS

### DISCUSSION THREAD

**Discussion by gramian** [ArcadeData/arcadedb#1057] [Ideas]
Dear Developers,

I would like to propose to add a server setting enabling [Cross Origin Resource Sharing](https://en.wikipedia.org/wiki/Cross-origin_resource_sharing) (CORS).

While publicly exposing a database server in entirety is generally a bad idea, just forwarding the `query` endpoint for a read-only user could be acceptable. Also during development and testing of a wrapping application or driver, such a setting could be useful.

As ArcadeDB uses Undertow, enabling CORS is dicussed here: https://stackoverflow.com/questions/42066845/how-to-enable-access-control-allow-origin-in-undertow

Best

**Comment by lvca** [comment]
I like this. If this can be solved by just adding this piece of code:

```java
httpServerExchange.getResponseHeaders()
                .put(new HttpString("Access-Control-Allow-Origin"), "*");
```

Then it's super easy, we could have a setting to enable/disable that. WDYT?

**Reply by gramian** [reply]
Of course, it turns out, this is more complicated: Since the server requires credentials, the CORS standard forbids to return `*` as origin. More formally: If `Access-Control-Allow-Credentials` is used (which is, due to the authorization), then `Access-Control-Allow-Origin` has to contain the specific origin (domain), see: https://portswigger.net/web-security/cors#server-generated-acao-header-from-client-specified-origin-header .

**Reply by gramian** [reply]
Particularly, this means a so-called CORS Preflight Request via an `OPTIONS` method has to succeed, which is currently not supported.

**Comment by tolgaulas** [comment]
As this is a security matter and users may like to define their own
constraints, I suggest making the value * only if the user did not define
their own value.


Tolga Ulas

**Reply by lvca** [reply]
Should it be enabled by default with `*`? Or disabled by default for security? We could define a new setting `arcadedb.server.cors`, where:
- if not defined (NULL), then it's disabled
- otherwise, the value goes in the `Access-Control-Allow-Origin`. Typically "*", but the users can put whatever they want

**Reply by gramian** [reply]
IMHO This should be deactivated by default.

**Reply by tolgaulas** [reply]
I concur.

**Comment by lvca** [comment]
I've pushed a new branch `http_cors` with the work in progress to provide full CORS:

https://github.com/ArcadeData/arcadedb/commit/5cb1ef114fe131a9bd5590b4de1a51493f66ed49

@gramian was mentioning an OPTIONS call is still missing. What else should we provide? Feel free to fork and contribute to this branch.
